### ISSUE TITLE
Providing Lazy<> or Owned<> leads to an exception

### ISSUE THREAD

**Issue by mareklinka** [autofac/Autofac.Extras.Moq#10]
Stack trace:

```
Autofac.Core.DependencyResolutionException: An error occurred during the activation of a particular registration. See the inner exception for details. Registration: Activator = Object (DelegateActivator), Services = [System.Lazy`1[[AutofacMoqTest.UnitTest1+Dependency, AutofacMoqTest, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]]], Lifetime = Autofac.Core.Lifetime.CurrentScopeLifetime, Sharing = Shared, Ownership = OwnedByLifetimeScope ---> Can not create proxy for type System.Lazy`1[[AutofacMoqTest.UnitTest1+Dependency, AutofacMoqTest, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]] because type AutofacMoqTest.UnitTest1+Dependency is not accessible. Make it public, or internal and mark your assembly with [assembly: InternalsVisibleTo("DynamicProxyGenAssembly2, PublicKey=0024000004800000940000000602000000240000525341310004000001000100c547cac37abd99c8db225ef2f6c8a3602f3b3606cc9891605d02baa56104f4cfc0734aa39b93bf7852f7d9266654753cc297e7d2edfe0bac1cdcf9f717241550e0a7b191195b7667bb4f64bcb8e2121380fd1d9d46ad2d92d2d15605093924cceaf74c4861eff62abf69b9291ed0a340e113be11e6a7d3113e92484cf7045cc7")] attribute, because assembly System.Private.CoreLib is strong-named. (See inner exception for details.) ---> Castle.DynamicProxy.Generators.GeneratorException: Can not create proxy for type System.Lazy`1[[AutofacMoqTest.UnitTest1+Dependency, AutofacMoqTest, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]] because type AutofacMoqTest.UnitTest1+Dependency is not accessible. Make it public, or internal and mark your assembly with [assembly: InternalsVisibleTo("DynamicProxyGenAssembly2, PublicKey=0024000004800000940000000602000000240000525341310004000001000100c547cac37abd99c8db225ef2f6c8a3602f3b3606cc9891605d02baa56104f4cfc0734aa39b93bf7852f7d9266654753cc297e7d2edfe0bac1cdcf9f717241550e0a7b191195b7667bb4f64bcb8e2121380fd1d9d46ad2d92d2d15605093924cceaf74c4861eff62abf69b9291ed0a340e113be11e6a7d3113e92484cf7045cc7")] attribute, because assembly System.Private.CoreLib is strong-named.
   at Castle.DynamicProxy.DefaultProxyBuilder.AssertValidTypeForTarget(Type type, Type target)
   at Castle.DynamicProxy.DefaultProxyBuilder.AssertValidTypeForTarget(Type type, Type target)
   at Castle.DynamicProxy.DefaultProxyBuilder.CreateClassProxyType(Type classToProxy, Type[] additionalInterfacesToProxy, ProxyGenerationOptions options)
   at Castle.DynamicProxy.ProxyGenerator.CreateClassProxy(Type classToProxy, Type[] additionalInterfacesToProxy, ProxyGenerationOptions options, Object[] constructorArguments, IInterceptor[] interceptors)
   at Moq.Proxy.CastleProxyFactory.CreateProxy(Type mockType, ICallInterceptor interceptor, Type[] interfaces, Object[] arguments)
   at Moq.Mock`1.<InitializeInstance>b__19_0()
   at Moq.Mock`1.OnGetObject()
   at Autofac.Extras.Moq.MoqRegistrationHandler.CreateMock(IComponentContext context, TypedService typedService)
   at Autofac.Builder.RegistrationBuilder.<>c__DisplayClass0_0`1.<ForDelegate>b__0(IComponentContext c, IEnumerable`1 p)
   at Autofac.Core.Activators.Delegate.DelegateActivator.ActivateInstance(IComponentContext context, IEnumerable`1 parameters)
   at Autofac.Core.Resolving.InstanceLookup.Activate(IEnumerable`1 parameters)
   --- End of inner exception stack trace ---
   at Autofac.Core.Resolving.InstanceLookup.Activate(IEnumerable`1 parameters)
   at Autofac.Core.Lifetime.LifetimeScope.GetOrCreateAndShare(Guid id, Func`1 creator)
   at Autofac.Core.Resolving.InstanceLookup.Execute()
   at Autofac.Core.Resolving.ResolveOperation.GetOrCreateInstance(ISharingLifetimeScope currentOperationScope, IComponentRegistration registration, IEnumerable`1 parameters)
   at Autofac.Core.Resolving.ResolveOperation.Execute(IComponentRegistration registration, IEnumerable`1 parameters)
   at lambda_method(Closure )
   at AutofacMoqTest.UnitTest1.Test_Lazy() in d:\Documents\Visual Studio 2017\Projects\AutofacMoqTest\AutofacMoqTest\UnitTest1.cs:line 18
```

Using the latest beta package, version `4.1.0-rc5-246`, the project is targeting .NET Core 2.0.

Simple repro (xUnit test project targeting Core 2.0):

``` c#
public class UnitTest1
{
    [Fact]
    public void Test_Lazy()
    {
        using (var mock = AutoMock.GetLoose())
        {
            mock.Provide(new Dependency(5));

            var sut = mock.Create<LazyTest>();
            var result = sut.DependencyFactory();

            Assert.NotNull(result);
        }
    }

    [Fact]
    public void Test_Owned()
    {
        using (var mock = AutoMock.GetLoose())
        {
            mock.Provide(new Dependency(5));

            var sut = mock.Create<OwnedTest>();
            var result = sut.DependencyFactory();

            Assert.NotNull(result);
        }
    }

    class LazyTest
    {
        public Func<Lazy<Dependency>> DependencyFactory { get; }

        public LazyTest(Func<Lazy<Dependency>> dependencyFactory)
        {
            this.DependencyFactory = dependencyFactory;
        }
    }

    class OwnedTest
    {
        public Func<Owned<Dependency>> DependencyFactory { get; }

        public OwnedTest(Func<Owned<Dependency>> dependencyFactory)
        {
            this.DependencyFactory = dependencyFactory;
        }
    }

    class Dependency
    {
        private readonly int parameter;

        public Dependency(int parameter)
        {
            this.parameter = parameter;
        }
    }
}
```

This used to work with package version 4.0.0 targeting .NET Framework 4.7.

I also have a forked version that seems to fix the issue, should open a PR soon.

**Comment by tillig** [comment]
Can you provide a little more info here?

**Comment by mareklinka** [comment]
Yes, sorry, submitted the issue early by accident. Updated.

**Comment by mareklinka** [comment]
I should also add that I'm using Autofac 4.6.1 and the Moq 4.7.127.

**Comment by tillig** [comment]
Duplicate of #3.

**Comment by tillig** [comment]
Fix released in 4.1.0 to NuGet.
